### PR TITLE
Action of Mass Matrix using Partial Assembly 

### PR DESCRIPTION
**Pull request by artv3**
Results with Seq, OpenMP, CUDA
```
Speedup Report (T_ref/T_var): ref var = Base_Seq  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,                                                                                   
Kernel         , Base_Seq , Lambda_Seq , RAJA_Seq , Base_OpenMP , Lambda_OpenMP , RAJA_OpenMP , Base_CUDA , Lambda_CUDA , RAJA_CUDA , RAJA_WORKGROUP_CUDA        
Apps_MASS3DPA  ,    1.000 ,    Not run ,    0.984 ,      42.143 ,       Not run ,      40.221 ,   178.824 ,     Not run ,   182.615 ,             Not run   
```

### CODE REVIEW

File: src/apps/MASS3DPA-OMP.cpp
Comment (rhornung67) [inline src/apps/MASS3DPA-OMP.cpp:18]: This #if defined line can be removed. It's redundant with the one inside the method below and that method needs to be seen during compilation.

**Review by rhornung67 — CHANGES_REQUESTED** [review]
@artv3 looks good. One minor comment about the #if defined line in the OpenMP file that should be addressed before merging.

File: src/apps/MASS3DPA-Seq.cpp
Comment (rhornung67) [inline src/apps/MASS3DPA-Seq.cpp:82]: Are these macro guarded code sections here for consistency with the GPU variants?

File: src/apps/MASS3DPA-Seq.cpp
Comment (rhornung67) [inline src/apps/MASS3DPA-Seq.cpp:115]: I'm also confused by the "threads" terminology in this sequential file. Would it be clearer to name the policies to better align with how the kernel variants in the files will execute?

File: src/apps/MASS3DPA.cpp
Comment (rhornung67) [inline src/apps/MASS3DPA.cpp:29]: So I picked the label for this RAJA feature based on its name in the code. Would you rather call it "teams"?

File: src/apps/MASS3DPA.hpp
Comment (rhornung67) [inline src/apps/MASS3DPA.hpp:78]: where is this "unroll" macro defined?

File: src/apps/MASS3DPA.hpp
Comment (rhornung67) [inline src/apps/MASS3DPA.hpp:235]: Have you tried compiling your branch with OpenMP target enabled? I think the runOpenMPTargetVariant method must be defined (it can be empty) here for the code to compile since it is pure virtual in KernelBase class.

**Review by rhornung67 — COMMENTED** [review]
A few questions and comments to address, double-check.

File: src/apps/MASS3DPA-Seq.cpp
Comment (artv3) [reply src/apps/MASS3DPA-Seq.cpp:115]: That is a good point; In general I was going for an abstraction of a hierarchical nested loop model that "fits" the block/thread model of GPU computing. I suppose we can refer the policies as "inner_i/outer_i" policies like that they are universal across programming models .

File: src/apps/MASS3DPA-Seq.cpp
Comment (artv3) [reply src/apps/MASS3DPA-Seq.cpp:82]: It's in case we compile with a device; at the moment we force having to provide both (host/device) policies. Future work in Launch/Teams will rework the framework so its not forced.

File: src/apps/MASS3DPA.hpp
Comment (artv3) [reply src/apps/MASS3DPA.hpp:78]: Its defined in the back-end specific variants. Depending on which back-end we use; unrolling loops can help or hurt performance; example CUDA/HIP so we either do or do not unroll.

File: src/apps/MASS3DPA.hpp
Comment (artv3) [reply src/apps/MASS3DPA.hpp:235]: I'll check once I get access to the a machine.

File: src/apps/MASS3DPA-Seq.cpp
Comment (rhornung67) [reply src/apps/MASS3DPA-Seq.cpp:82]: Please add a GitHub issue to address that in the kernel implementation when it's supported in RAJA.

File: src/apps/MASS3DPA.cpp
Comment (artv3) [reply src/apps/MASS3DPA.cpp:29]: I think we should go with teams, it's been adopted colloquially and reminds folks about computing with hierarchical parallelism. I can make that change here.

File: src/apps/MASS3DPA.hpp
Comment (artv3) [reply src/apps/MASS3DPA.hpp:236]: Good point, I just pushed some fixes. It also seems it needs the -OpenMP.cpp file. I wasn't able to get the code to build without it.

**Comment by artv3** [comment]
@rhornung67 @MrBurmark , can I trouble you for another review? I can update mat mat example after we merge this PR in.

**Comment by rhornung67** [comment]
@artv3 Jason and I are hashing something things out about perf suite organization. We have several PRs coming soon. Do you mind holding off on your PRs until we have things mostly sorted out? Hopefuylly, this will prevent you having to make too many changes in your work.

**Comment by artv3** [comment]
> @artv3 Jason and I are hashing something things out about perf suite organization. We have several PRs coming soon. Do you mind holding off on your PRs until we have things mostly sorted out? Hopefuylly, this will prevent you having to make too many changes in your work.

That is fine; just let me know when we are ready.

File: src/apps/MASS3DPA-Cuda.cpp
Comment (MrBurmark) [inline src/apps/MASS3DPA-Cuda.cpp:50]: This spacing/indentation seems weird

Comment (artv3) [reply src/apps/MASS3DPA-Cuda.cpp]: fixed

File: src/apps/MASS3DPA.cpp
Comment (MrBurmark) [inline src/apps/MASS3DPA.cpp:31]: guard against getting 0, when problem size is less than Q1D*Q1D

File: src/apps/MASS3DPA.cpp
Comment (MrBurmark) [inline src/apps/MASS3DPA.cpp:29]: Default size is the default problem size, they should be defined consistently

File: src/apps/MASS3DPA.hpp
Comment (artv3) [inline src/apps/MASS3DPA.hpp:14]: Copy and paste the code.

File: src/apps/MASS3DPA.cpp
Comment (artv3) [reply src/apps/MASS3DPA.cpp:31]: move to line 28.

File: src/apps/MASS3DPA.cpp
Comment (artv3) [inline src/apps/MASS3DPA.cpp:26]: m_NE_default*Q1D*Q1D

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**
